File: MarkdownEditor/MarkdownEditor/Services/FileManagerService.swift
import Foundation

final class FileManagerService {
    static let shared = FileManagerService()
    private let fm = FileManager.default
    private init() {}

    /// Build a file tree from a local repo directory
    func buildFileTree(at rootURL: URL) -> [FileNode] {
        buildFileTree(at: rootURL, repoRoot: rootURL)
    }

    private func buildFileTree(at directoryURL: URL, repoRoot: URL) -> [FileNode] {
        guard let contents = try? fm.contentsOfDirectory(
            at: directoryURL,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: []
        ) else {
            return []
        }

        let rootPath = repoRoot.standardizedFileURL.path
        var nodes: [FileNode] = []

        for url in contents {
            let name = url.lastPathComponent

            // Skip metadata and originals
            if name == ".repo-metadata.json" || name == ".originals" { continue }

            let isDirectory = (try? url.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory ?? false
            let filePath = url.standardizedFileURL.path
            var relativePath = filePath
            if filePath.hasPrefix(rootPath) {
                relativePath = String(filePath.dropFirst(rootPath.count))
                if relativePath.hasPrefix("/") {
                    relativePath.removeFirst()
                }
            }

            if isDirectory {
                let children = buildFileTree(at: url, repoRoot: repoRoot)
                let node = FileNode(
                    name: name,
                    path: relativePath,
                    isDirectory: true,
                    fileType: .binary,
                    children: children
                )
                nodes.append(node)
            } else {
                let fileType = FileTypeDetector.detect(filename: name)
                let node = FileNode(
                    name: name,
                    path: relativePath,
                    isDirectory: false,
                    fileType: fileType
                )
                nodes.append(node)
            }
        }

        return nodes.sorted()
    }

    /// Build a file tree annotated with change status from a list of FileChanges.
    func buildFileTree(at rootURL: URL, changes: [FileChange]) -> [FileNode] {
        let changeMap = Dictionary(uniqueKeysWithValues: changes.map { ($0.path, $0.changeType) })
        var nodes = buildFileTree(at: rootURL)
        annotateNodes(&nodes, changeMap: changeMap)
        return nodes
    }

    /// Recursively annotate nodes with their change type.
    private func annotateNodes(_ nodes: inout [FileNode], changeMap: [String: FileChangeType]) {
        for i in nodes.indices {
            if let changeType = changeMap[nodes[i].path] {
                nodes[i].changeType = changeType
            }
            if var children = nodes[i].children {
                annotateNodes(&children, changeMap: changeMap)
                nodes[i].children = children
            }
        }
    }

    /// Read file content as string
    func readFileContent(at url: URL) -> String? {
        try? String(contentsOf: url, encoding: .utf8)
    }

    /// Write file content
    func writeFileContent(_ content: String, to url: URL) throws {
        try content.write(to: url, atomically: true, encoding: .utf8)
    }

    /// Create a new file
    func createFile(named name: String, in directoryURL: URL) throws -> URL {
        let fileURL = directoryURL.appendingPathComponent(name)
        fm.createFile(atPath: fileURL.path, contents: nil)
        return fileURL
    }

    /// Create a new directory
    func createDirectory(named name: String, in parentURL: URL) throws -> URL {
        let dirURL = parentURL.appendingPathComponent(name)
        try fm.createDirectory(at: dirURL, withIntermediateDirectories: true)
        return dirURL
    }

    /// Delete a file or directory
    func delete(at url: URL) throws {
        try fm.removeItem(at: url)
    }

    /// Rename a file or directory
    func rename(at url: URL, to newName: String) throws -> URL {
        let newURL = url.deletingLastPathComponent().appendingPathComponent(newName)
        try fm.moveItem(at: url, to: newURL)
        return newURL
    }

    /// Get file size
    func fileSize(at url: URL) -> Int64 {
        let attributes = try? fm.attributesOfItem(atPath: url.path)
        return attributes?[.size] as? Int64 ?? 0
    }

    /// Check if path is a text file that can be edited
    func isEditable(at url: URL) -> Bool {
        let fileType = FileTypeDetector.detect(filename: url.lastPathComponent)
        return fileType == .markdown || fileType == .text
    }
}

File: android/lib/services/file_manager_service.dart
import 'dart:io';
import '../models/file_node.dart';
import '../models/git_status.dart';

class FileManagerService {
  static final FileManagerService shared = FileManagerService._();
  FileManagerService._();

  static const _skipNames = {'.repo-metadata.json', '.originals'};

  List<FileNode> buildFileTree(String rootPath) {
    return _buildTree(Directory(rootPath), rootPath);
  }

  List<FileNode> buildFileTreeWithChanges(
    String rootPath,
    List<FileChange> changes,
  ) {
    final changeMap = {for (final c in changes) c.path: c.changeType};
    var nodes = buildFileTree(rootPath);
    _annotateNodes(nodes, changeMap);
    return nodes;
  }

  List<FileNode> _buildTree(Directory dir, String rootPath) {
    final List<FileNode> nodes = [];

    final List<FileSystemEntity> entries;
    try {
      entries = dir.listSync();
    } catch (_) {
      return nodes;
    }

    for (final entry in entries) {
      final name = entry.path.split('/').last;
      if (_skipNames.contains(name)) continue;

      final relPath = relativePath(entry.path, rootPath);

      if (entry is Directory) {
        final children = _buildTree(entry, rootPath);
        nodes.add(FileNode(
          name: name,
          path: relPath,
          isDirectory: true,
          fileType: FileType.binary,
          children: children,
        ));
      } else if (entry is File) {
        final fileType = FileTypeDetector.detect(name);
        nodes.add(FileNode(
          name: name,
          path: relPath,
          isDirectory: false,
          fileType: fileType,
        ));
      }
    }

    nodes.sort();
    return nodes;
  }

  void _annotateNodes(List<FileNode> nodes, Map<String, FileChangeType> changeMap) {
    for (final node in nodes) {
      if (changeMap.containsKey(node.path)) {
        node.changeType = changeMap[node.path];
      }
      if (node.children != null) {
        _annotateNodes(node.children!, changeMap);
      }
    }
  }

  String? readFileContent(String path) {
    try {
      return File(path).readAsStringSync();
    } catch (_) {
      return null;
    }
  }

  void writeFileContent(String content, String path) {
    File(path).writeAsStringSync(content, flush: true);
  }

  String createFile(String name, String directoryPath) {
    final filePath = '$directoryPath/$name';
    File(filePath).createSync(recursive: true);
    return filePath;
  }

  String createDirectory(String name, String parentPath) {
    final dirPath = '$parentPath/$name';
    Directory(dirPath).createSync(recursive: true);
    return dirPath;
  }

  void delete(String path) {
    final entity = FileSystemEntity.typeSync(path) == FileSystemEntityType.directory
        ? Directory(path)
        : File(path) as FileSystemEntity;
    entity.deleteSync(recursive: true);
  }

  void move(String sourcePath, String destPath) {
    FileSystemEntity.typeSync(sourcePath) == FileSystemEntityType.directory
        ? Directory(sourcePath).renameSync(destPath)
        : File(sourcePath).renameSync(destPath);
  }

  String rename(String path, String newName) {
    final parent = path.substring(0, path.lastIndexOf('/'));
    final newPath = '$parent/$newName';
    move(path, newPath);
    return newPath;
  }

  int fileSize(String path) {
    try {
      return File(path).statSync().size;
    } catch (_) {
      return 0;
    }
  }

  bool isEditable(String filename) {
    final type = FileTypeDetector.detect(filename);
    return type == FileType.markdown || type == FileType.text;
  }

  static String relativePath(String fullPath, String rootPath) {
    if (fullPath.startsWith(rootPath)) {
      var rel = fullPath.substring(rootPath.length);
      if (rel.startsWith('/')) rel = rel.substring(1);
      return rel;
    }
    return fullPath;
  }

  // Directory size helper
  int directorySize(String path) {
    try {
      int size = 0;
      Directory(path)
          .listSync(recursive: true)
          .whereType<File>()
          .forEach((f) => size += f.statSync().size);
      return size;
    } catch (_) {
      return 0;
    }
  }
}
